File: src/main/java/no/fintlabs/application/crd/FlaisApplicationSpec.java
package no.fintlabs.application.crd;

import com.fasterxml.jackson.annotation.JsonPropertyDescription;
import io.fabric8.kubernetes.api.model.EnvFromSource;
import io.fabric8.kubernetes.api.model.EnvVar;
import io.fabric8.kubernetes.api.model.ResourceRequirements;
import io.fabric8.kubernetes.api.model.apps.DeploymentStrategy;
import lombok.*;
import no.fintlabs.FlaisSpec;
import no.fintlabs.onepassword.OnePassword;

import java.util.ArrayList;
import java.util.List;

@Getter
@Setter
//@Builder
@NoArgsConstructor
@AllArgsConstructor
public class FlaisApplicationSpec implements FlaisSpec {

    @JsonPropertyDescription("If present this application is consider to be a county application.")
    private String orgId;

    private List<EnvVar> env = new ArrayList<>();

    private List<EnvFromSource> envFrom = new ArrayList<>();

    private String image;

    private Integer replicas;

    private ResourceRequirements resources;

    private Integer port = 8080;

    private String restartPolicy = "Always";

    private DeploymentStrategy strategy;

    private Prometheus prometheus = new Prometheus();

    private OnePassword onePassword = new OnePassword();

}

File: src/main/kotlin/no/fintlabs/application/api/v1alpha1/FlaisApplicationSpec.kt
package no.fintlabs.application.api.v1alpha1

import io.fabric8.generator.annotation.Min
import io.fabric8.generator.annotation.ValidationRule
import io.fabric8.kubernetes.api.model.EnvFromSource
import io.fabric8.kubernetes.api.model.EnvVar
import io.fabric8.kubernetes.api.model.Quantity
import io.fabric8.kubernetes.api.model.ResourceRequirements
import io.fabric8.kubernetes.api.model.ResourceRequirementsBuilder
import io.fabric8.kubernetes.api.model.apps.DeploymentStrategy
import no.fintlabs.common.WithKafka
import no.fintlabs.common.WithOnePassword
import no.fintlabs.common.WithPostgres
import no.fintlabs.common.api.v1alpha1.Database
import no.fintlabs.common.api.v1alpha1.FlaisResourceSpec
import no.fintlabs.common.api.v1alpha1.Kafka
import no.fintlabs.common.api.v1alpha1.OnePassword
import no.fintlabs.common.api.v1alpha1.Probes

data class FlaisApplicationSpec(
    override val orgId: String = "",
    @get:Min(0.0)
    val replicas: Int = 1,
    override val image: String = "",
    override val imagePullPolicy: String? = null,
    override val imagePullSecrets: List<String> = emptyList(),
    override val env: List<EnvVar> = emptyList(),
    override val envFrom: List<EnvFromSource> = emptyList(),
    override val resources: ResourceRequirements =
        ResourceRequirementsBuilder()
            .addToRequests("cpu", Quantity("250m"))
            .addToRequests("memory", Quantity("256Mi"))
            .addToLimits("cpu", Quantity("500m"))
            .addToLimits("memory", Quantity("512Mi"))
            .build(),
    val probes: Probes? = null,
    @Min(1.0) val port: Int = 8080,
    @Deprecated("Does not exist on when applied on DeploymentSpec")
    val restartPolicy: String = "Always",
    val strategy: DeploymentStrategy? = null,
    override val onePassword: OnePassword? = null,
    override val kafka: Kafka = Kafka(),
    override val database: Database = Database(),
    val prometheus: Metrics = Metrics(),
    val url: Url = Url(),
    val ingress: Ingress? = null,
    override val observability: ApplicationObservability? = null,
) : FlaisResourceSpec, WithOnePassword, WithKafka, WithPostgres
